<a href="https://colab.research.google.com/github/sdasmaha/Ml_Engineers_445/blob/master/Exercise_3_Submission_DasMahapatra.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Notebook Excercise #3
#Srinjoy DasMahapatra

IDE => Google Colab Pro

In [ ]:
import numpy as np
import pandas as pd
from scipy import stats as sps
from matplotlib import pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from IPython.display import display, Latex

# 1. Non-centered Data and Principal Component Analysis (PCA)

## 1.1

In [ ]:
A = np.random.randn(3,2)
c = np.random.randn(3,1)
c1 = np.random.randn(3,1)

## 1.2

In [ ]:
B = np.random.randn(2,250)
X_1 = np.matmul(A,B)
X = np.add(X_1,c)

In [ ]:
np.shape(X)

(3, 250)

## 1.3

In [ ]:
np.linalg.matrix_rank(X, tol=None)

3

## 1.4

The choices that justify when c != 0 gurantees that the data remains centered are:




## 1.5

### (a)

### (b)

In [ ]:
U, s, Vh = np.linalg.svd(X)

In [ ]:
U = A / np.linalg.norm(A)
U

array([[-0.27895853, -0.44608336],
       [ 0.24112474,  0.56748347],
       [ 0.39693637, -0.43064447]])

In [ ]:
U = Vh.T[:,0:2]


[[-8.72668286e-02  4.30697386e-02]
 [-9.43144968e-02  5.40067097e-02]
 [-6.73219668e-02  4.64344070e-02]
 [-2.64101880e-02 -4.68493055e-02]
 [ 2.00101550e-02  3.75747198e-02]
 [-5.01643451e-02 -7.64296269e-02]
 [-9.30647891e-02  6.92263170e-02]
 [ 1.07801199e-02 -7.39058163e-02]
 [-5.45542229e-02 -4.77600257e-02]
 [-1.63901731e-02 -1.97505437e-03]
 [-4.00214960e-02 -1.53916668e-02]
 [-4.52076970e-02 -3.72897253e-02]
 [-3.30917066e-02  2.95722891e-02]
 [-3.92086536e-02 -3.95401810e-02]
 [-1.69120571e-02  1.54340739e-02]
 [-3.28398408e-02  5.81257561e-03]
 [-1.22474054e-01  1.57304078e-01]
 [-5.65465220e-02 -1.18603845e-02]
 [-2.30015043e-02  1.59652889e-03]
 [-2.60161079e-02  1.39020031e-03]
 [-2.26839657e-02 -1.18746758e-01]
 [-9.94895493e-02  1.80089801e-02]
 [-2.98195251e-02  8.76433903e-02]
 [-3.52328893e-02  1.40859030e-02]
 [-9.31681457e-02  1.02119797e-02]
 [-4.87071756e-02 -3.29664071e-02]
 [-7.40328658e-02 -1.48466771e-01]
 [ 2.14899228e-03  3.28029472e-02]
 [-5.43676410e-02  4

In [ ]:
s

array([29.99301282, 19.70716817,  3.0348222 ])

In [ ]:
U = U[:,0:2]
U

array([[-0.32239499,  0.04649212],
       [-0.40860728, -0.90778491],
       [-0.85387444,  0.41685133]])

In [ ]:
U_UT = np.matmul(U,U.T)
U_UT

array([[ 0.10610004,  0.08952809,  0.29466514],
       [ 0.08952809,  0.99103336, -0.02951204],
       [ 0.29466514, -0.02951204,  0.9028666 ]])

In [ ]:
np.shape(X)

(3, 250)

In [ ]:
X_Hat = np.matmul(U_UT,X)

In [ ]:
np.shape(X)

(3, 250)

### (c)

In [ ]:
print(np.linalg.norm(X_Hat-X)**2)


9.210145755649878


## 1.6

In [ ]:
m = (np.mean(X_1, axis=0))

In [ ]:
X_1_Hat =  (U_UT@(X_1-m))+m

In [ ]:
print(np.linalg.norm(X_1_Hat-X)**2)


1842.9953043754551


# 2. Preprocessing (Centering) and PCA Using `scikit learn`

## 2.1

In [ ]:
# your code for 2.1

## 2.2

In [ ]:
# your code for 2.2

## 2.3

In [ ]:
# your code for 2.3

## 2.4

In [ ]:
# your code for 2.4

## 2.5

In [ ]:
# your code for 2.5

## 2.6

In [ ]:
# your code for 2.6

## 2.7

In [ ]:
# your code for 2.7

# 3. PCA as a Denoising Tool

## 3.1

In [ ]:
# your code for 3.1

## 3.2

In [ ]:
# your code for 3.2

## 3.3

In [ ]:
# your code for 3.3

## 3.4

In [ ]:
# your code for 3.4

## 3.5

In [ ]:
# your code for 3.5

## 3.6

In [ ]:
# your code for 3.6

## 3.7

In [ ]:
# your code for 3.7

## 3.8

_Provide your answer for 3.8 in this markdown cell_